### PR TITLE
Icinga DB: dump after all connections connected

### PR DESCRIPTION
**Pull request by Al2Klimov**
... no to loose any messages.

### CODE REVIEW

**Comment by Al2Klimov** [comment]
## Blocked by

* #8933

**Comment by Al2Klimov** [comment]
## Works

```
[2021-07-27 18:13:23 +0200] information/IcingaDB: 'icingadb' started.
[2021-07-27 18:13:23 +0200] information/ConfigItem: Activated all objects.
[2021-07-27 18:13:23 +0200] information/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/ApiListener: Updating object authority for local objects.
[2021-07-27 18:13:23 +0200] information/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6379'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Connected to Redis server
[2021-07-27 18:13:23 +0200] notice/WorkQueue: Spawning WorkQueue threads for 'IcingaDB'
[2021-07-27 18:13:23 +0200] notice/IcingaDB: Firing and forgetting query: 'XADD' 'icinga:stats' 'MAXLEN' '1' '*' 'ApiListener' '{"perfdata":[],"status":{}}' ...
[2021-07-27 18:13:23 +0200] information/IcingaDB: Starting initial config/status dump
```

File: lib/icingadb/icingadb.cpp [hunk 1]
```diff
@@ -67,8 +69,23 @@ void IcingaDB::Start(bool runtimeCreated)
...
 		GetEnableTls(), GetInsecureNoverify(), GetCertPath(), GetKeyPath(), GetCaPath(), GetCrlPath(),
 		GetTlsProtocolmin(), GetCipherList(), GetDebugInfo());
-	m_Rcon->SetConnectedCallback([this](boost::asio::yield_context& yc) {
-		m_WorkQueue.Enqueue([this]() { OnConnectedHandler(); });
+
+	m_Rcon->SetConnectedCallback([this](yield_context& yc) {
+		m_Rcon->SetConnectedCallback({});
+		m_RconsPending.store(m_Rcons.size());
+
+		for (auto& kv : m_Rcons) {
```
Comment (julianbrost) [inline lib/icingadb/icingadb.cpp:77]: `m_Rcons` is only initialized after this connection is started, so this access could happen while that map is not yet fully initialized. Simply moving setting this callback and the call fo `Start()` after the loop should fix this.

**Review by julianbrost — CHANGES_REQUESTED** [review]
In general, I'd do as little as required within the callback. Initializing `m_RconsPending` and setting the callbacks of the child connections can also be done outside of the callback of the parent connection.

File: lib/icingadb/icingadb.cpp (hunk 1 above)
Comment (julianbrost) [reply lib/icingadb/icingadb.cpp:77]: (This was already introduced by #8933, somehow I've missed that there.)

Comment (Al2Klimov) [reply lib/icingadb/icingadb.cpp:77]: That’s #8932's purpose.

**Pull request closed without merging**